stmtools/stm.py

```diff
@@ -569,6 +669,82 @@ def _compute_morton_code(xx, yy):
     -------
     array_like
         An array with Morton codes per coordinate pair.
+
     """
     code = [pm.interleave(int(xi), int(yi)) for xi, yi in zip(xx, yy, strict=True)]
     return code
+
+
+def _enrich_from_raster_block(ds, dataraster, fields, method):
+    """Enrich the ds (SpaceTimeMatrix) from one or more fields of a raster dataset.
+
+    scipy is required. It uses xarray.Dataset.interp_like to interpolate the
+    raster dataset to the coordinates of ds.
+    https://docs.xarray.dev/en/stable/generated/xarray.Dataset.interp.html
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        SpaceTimeMatrix to enrich
+    dataraster : xarray.Dataset | xarray.DataArray
+        Input data for enrichment
+    fields : str or list of str
+        Field name(s) in the dataset for enrichment
+    method : str, optional
+        Method of interpolation, by default "nearest", see
+
+    Returns
+    -------
+    xarray.Dataset
+
+    """
+    # interpolate the raster dataset to the coordinates of ds
+    interpolated = dataraster.interp(ds.coords, method=method)
+
+    # Assign these values to the corresponding points in ds
+    for field in fields:
+        ds[field] = xr.DataArray(interpolated[field].data, dims=ds.dims, coords=ds.coords)
+    return ds
+
+
+def _enrich_from_points_block(ds, datapoints, fields):
+    """Enrich the ds (SpaceTimeMatrix) from one or more fields of a point dataset.
+
+    https://docs.xarray.dev/en/latest/generated/xarray.DataArray.sel.html#xarray.DataArray.sel
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        SpaceTimeMatrix to enrich
+    datapoints : xarray.Dataset | xarray.DataArray
+        Input data for enrichment
+    fields : str or list of str
+        Field name(s) in the dataset for enrichment
+
+    Returns
+    -------
+    xarray.Dataset
+
+    """
+    # unstak the dimensions
+    for dim in datapoints.dims:
+        if dim not in datapoints.coords:
+            indexer = {
+                dim: [
+                    coord for coord in datapoints.coords if dim in datapoints[coord].dims
+                    ]
+                }
+            datapoints = datapoints.set_index(indexer)
+            datapoints = datapoints.unstack(dim)
+
+    # do selection
+    indexers = {coord: ds[coord] for coord in list(datapoints.coords.keys())}
+    selections = datapoints.sel(indexers, method="nearest")
```

Comment:
I did a test using `sel` to get nearest time location, with `datetime64[ns]` format. This works under two conditions:

1. For `time` dimension, in a specific combination: `method=nearest` + `datetime64[ns]` format, Xarray requires the time coordinates of `ds` to be monotonic increasing or decreasing. 
2. For `lat` and `lon` dimension, the duplicated values should not exists in the corresponding coordinates.

Can we add check in the beginning of `enrich_from_dataset` to enforce this?

An example check for time:

```python
# This should be True
np.all(np.diff(ds['time'].values) > 0) or np.all(np.diff(ds['time'].values) < 0) 
```

An example check for lat and lon:
```python
# Following should be True
np.unique(ds['lat'].values).shape == ds['lat'].values.shape
np.unique(ds['lon'].values).shape == ds['lon'].values.shape
```

Reply:
good point. I added two [util functions](https://github.com/TUDelftGeodesy/stmtools/pull/66/commits/92d53016e43c990037cad934878a17f535ee45ed) to do the checks and call them in stem `_enrich_from_points_block` [here](https://github.com/TUDelftGeodesy/stmtools/pull/66/commits/b85415aedd0a582b2b956c3a2410b24023c647bd).

Reply:
I noticed that with KDTree, these two conditions are not needed. So I removed the checks from `_enrich_from_points_block`.

Reply:
I added some tests for these conditions

Reply:
@rogerkuou to check if the points are unique, the test `np.unique(ds['lat'].values).shape == ds['lat'].values.shape` is not enough because it only checks the duplicates in one dimension here `lat`. However, for example, points can be located on one line. 

Instead, we need a test if there are cases where `(lat, lon, time)` are duplicated. Functions like `xarray.Dataset.drop_duplicates` and `pandas.DataFrame.duplicated` can be used to write a test. But these functions only work on `dim` and not `coords`. In our cases, `lat` and `lon` are `coords` and `space` is the `dim`. So we might need to use `unstack` which leads to memory problems. 

As discussed, scipy `KDTree` works if coordinates `(lat, lon, time)` are duplicated and the values of variables e.g. temperature are the same too, I added a test for this. If the values of variables are not the same for duplicated coordinates, MacOs and linux behave differently to pick up a value related to the nearest neighbor. For this extreme case, I suggest creating a new issue and fixing it in another PR.